Implement history action user activation

lukewarlow opened the pull request:
#### 50773fa12cff9404aa56ad0bcac5f09e9ab6c194
<pre>
Implement history action user activation
<a href="https://bugs.webkit.org/show_bug.cgi?id=266554">https://bugs.webkit.org/show_bug.cgi?id=266554</a>

Reviewed by Alex Christensen.

This implements the history action user activation algorithm from the HTML spec.

Spec URL: <a href="https://html.spec.whatwg.org/multipage/interaction.html#history-action-activation">https://html.spec.whatwg.org/multipage/interaction.html#history-action-activation</a>

* LayoutTests/fast/html/history-action-user-activation-expected.txt: Added.
* LayoutTests/fast/html/history-action-user-activation.html: Added.
* LayoutTests/platform/gtk/TestExpectations:
* LayoutTests/platform/ios/TestExpectations:
* LayoutTests/platform/mac-wk1/TestExpectations:
* LayoutTests/platform/wincairo/TestExpectations:
* LayoutTests/platform/wpe/TestExpectations:
* Source/WebCore/page/LocalDOMWindow.cpp:
(WebCore::LocalDOMWindow::hasHistoryActionActivation const):
(WebCore::LocalDOMWindow::consumeHistoryActionUserActivation):
* Source/WebCore/page/LocalDOMWindow.h:
* Source/WebCore/testing/Internals.cpp:
(WebCore::Internals::hasHistoryActionActivation):
(WebCore::Internals::consumeHistoryActionUserActivation):
* Source/WebCore/testing/Internals.h:
* Source/WebCore/testing/Internals.idl:

Canonical link: <a href="https://commits.webkit.org/272590@main">https://commits.webkit.org/272590@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a07a129634a27991489ed0a7102f5b59733a7a4c

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/32223 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/10957 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/34038 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/34783 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/29202 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/33073 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/13310 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/8167 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/28770 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/32641 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/9237 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/28811 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/8049 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/8197 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/28729 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/36128 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/29297 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/29175 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/34326 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/8323 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/6260 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/32189 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/9964 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/28497 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/7526 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/8956 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/8862 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

achristensen07 commented on Source/WebCore/page/LocalDOMWindow.cpp:
```diff
@@ -1588,6 +1588,14 @@ bool LocalDOMWindow::hasStickyActivation() const
+// When the last history-action activation timestamp of W is not equal to the last activation timestamp of W,
+// then W is said to have history-action activation.
+// (https://html.spec.whatwg.org/multipage/interaction.html#history-action-activation)
+bool LocalDOMWindow::hasHistoryActionActivation() const
```
These two functions are currently only reachable by test code.  Their references to the HTML spec indicates that they should actually be hooked up to user interaction code or something like that.

lukewarlow replied:
It will be used in follow up patches. The close watcher API as well as the navigation API relies on this.

achristensen07 commented on Source/WebCore/page/LocalDOMWindow.cpp:
```diff
@@ -1611,6 +1619,23 @@ void LocalDOMWindow::consumeLastActivationIfNecessary()
+    if (!hasHistoryActionActivation())
+        return false;
+
+    for (RefPtr<Frame> frame = this->frame() ? &this->frame()->tree().top() : nullptr; frame; frame = frame->tree().traverseNext()) {
```
Do we really want to traverse ALL frames?  Does this mean that an iframe will affect its parent frame, even if they are from different origins?

lukewarlow replied:
This implementation matches that of consumeTransientActivation.

https://html.spec.whatwg.org/multipage/interaction.html#consume-history-action-user-activation shows that the two algorithms are almost identical.

achristensen07 replied:
That may be true, but it is implemented in a way that is only reachable by test code.  That algorithm is intended to be useable by the web platform.

lukewarlow replied:
Apologies I don't understand what the issue with them is? 

https://github.com/WebKit/WebKit/blob/8974c3b7c78a830bf62cdb2a5b8384f73e35a696/Source/WebCore/html/closewatcher/CloseWatcher.cpp#L105 here's the line in a follow up patch where I use these

achristensen07 replied:
Oh, the comment "That may be true..." was meant for the other thread about only being reachable from test code.  I see your plan a bit better now.

achristensen07 commented on Source/WebCore/page/LocalDOMWindow.h:
```diff
@@ -495,6 +497,10 @@ class LocalDOMWindow final
+    // For the purpose of tracking user activation, each Window W has a last history-action activation timestamp. This is a number indicating the last time W got
+    // an activation notification. It corresponds to a DOMHighResTimeStamp value except for one case: positive infinity indicates that W has never
+    // been activated. The initial value is positive infinity.
+    MonotonicTime m_lastHistoryActionActivationTimestamp { MonotonicTime::infinity() };
```
This comment is a bit verbose.  The initial value is clear from the code.

lukewarlow replied:
Fair enough will remove as part of any required changes

lukewarlow replied:
Removed

achristensen07 reviewed (approved):
Approved.

lukewarlow committed:
Implement history action user activation
https://bugs.webkit.org/show_bug.cgi?id=266554

Reviewed by Alex Christensen.

This implements the history action user activation algorithm from the HTML spec.

Spec URL: https://html.spec.whatwg.org/multipage/interaction.html#history-action-activation

* LayoutTests/fast/html/history-action-user-activation-expected.txt: Added.
* LayoutTests/fast/html/history-action-user-activation.html: Added.
* LayoutTests/platform/gtk/TestExpectations:
* LayoutTests/platform/ios/TestExpectations:
* LayoutTests/platform/mac-wk1/TestExpectations:
* LayoutTests/platform/wincairo/TestExpectations:
* LayoutTests/platform/wpe/TestExpectations:
* Source/WebCore/page/LocalDOMWindow.cpp:
(WebCore::LocalDOMWindow::hasHistoryActionActivation const):
(WebCore::LocalDOMWindow::consumeHistoryActionUserActivation):
* Source/WebCore/page/LocalDOMWindow.h:
* Source/WebCore/testing/Internals.cpp:
(WebCore::Internals::hasHistoryActionActivation):
(WebCore::Internals::consumeHistoryActionUserActivation):
* Source/WebCore/testing/Internals.h:
* Source/WebCore/testing/Internals.idl:

Canonical link: https://commits.webkit.org/272590@main
